### PR TITLE
Refactor model - MLM and CLM

### PR DESCRIPTION
**Pull request by gonzalobenegas**
- Refactors inference code to allow coupling data transforms and model scoring functions
- Adds support for autoregressive/causal language model (CLM), including Marin checkpoints (`examples/marin_evolutionary_constraint`)

### CODE REVIEW

**Comment by gonzalobenegas** [comment]
@eric-czech I wanted to get some feedback from you on the overall design. I'm planning on taking care of details later (e.g. shape suffixes, asserts).

File: biofoundation/model.py [hunk 1]
```diff
@@ -1,60 +1,115 @@
...
 
-class RefLogProbMLM(nn.Module):
-    """Reference Log Probability Masked Language Model wrapper.
+class HFLanguageModel(LanguageModel):
+    def __init__(self, model: PreTrainedModel):
+        super().__init__()
+        self.model = model
+
+    def forward(self, input_ids: Int[Tensor, "B L"]) -> Float[Tensor, "B L V"]:
+        return cast(Float[Tensor, "B L V"], self.model(input_ids).logits)
```
Comment (gonzalobenegas) [inline biofoundation/model.py:27]: This is an adapter for DNA language models, so it can handle both HF models but also other important ones such as Evo that have a slightly different API.

File: examples/marin_evolutionary_constraint.py
Comment (gonzalobenegas) [inline examples/marin_evolutionary_constraint.py:1]: Not requesting feedback, just sharing that I trained a Marin model for 5k steps and saw consistent improvements on this eval.

File: biofoundation/model.py (hunk 1 above)
Comment (gonzalobenegas) [reply biofoundation/model.py:27]: I'm torn about going one step further and have MLM and CLM abstract classes each implementing a `LanguageModel` abstract method `reflogprob` (currently this logic is in functions `compute_reflogprob_mlm` and `compute_reflogprob_clm`), to prevent the user from accidentally running the MLM scoring function on a CLM such as Evo.

Comment (gonzalobenegas) [reply biofoundation/model.py:27]: Also, besides the batch-level `reflogprob`, there could be a dataset-level `reflogprob_on_dataset(self, dataset)`.

File: biofoundation/inference.py [hunk 2]
```diff
@@ -1,11 +1,59 @@
...
+    run_inference,
+    compute_fn=compute_reflogprob_mlm,
+    data_transform_fn=transform_reflogprob_mlm,
+)
+
+run_reflogprob_clm = partial(
+    run_inference,
+    compute_fn=compute_reflogprob_clm,
+    data_transform_fn=transform_reflogprob_clm,
+)
```
[the remark this answers is not in the document]
Comment (gonzalobenegas) [reply biofoundation/inference.py:53]: I'm also wondering how to handle the averaging with predictions from reverse complement.

File: biofoundation/model.py (hunk 1 above)
Comment (gonzalobenegas) [reply biofoundation/model.py:27]: Thanks, this is very helpful. I like the idea of Protocols.

I'm thinking on how we could still prevent running MLM scoring function on a CLM. One challenge is that both MLM and CLM return logits of exactly the same type, even though they carry semantically different information. So if both MLM and CLM protocols are specified by `forward` or `get_logits` with the same shape, there's no way to distinguish them based on structural subtyping, right?

I was wondering if a good idea would be to just change the name of the method. e.g. a MLM protocol requiring `get_logits_mlm` and a CLM protocol requiring `get_logits_clm`.

Comment (gonzalobenegas) [reply biofoundation/model.py:27]: I'm leaning towards (3). I think there might be some issues with (1):
- When you do `model = AutoModelForCausalLM.from_pretrained(...)`, `model` is no longer an `AutoModelForCausalLM`, so it's hard to enforce typing:
```python
model = AutoModelForCausalLM.from_pretrained("LongSafari/hyenadna-tiny-1k-seqlen-hf", trust_remote_code=True)

def print_class_hierarchy(obj, indent=0):
    print("  " * indent + obj.__name__)
    for base in obj.__bases__:
        print_class_hierarchy(base, indent + 1)

print_class_hierarchy(type(model))
#HyenaDNAForCausalLM
#  HyenaDNAPreTrainedModel
#    PreTrainedModel
#      Module
#        object
#      EmbeddingAccessMixin
#        object
#      ModuleUtilsMixin
#        object
#      PushToHubMixin
#        object
#      PeftAdapterMixin
#        object
```
- I also think it's quite a lot of boilerplate to add, you need to define a couple of classes and register them in the Auto class mapping.

Comment (eric-czech) [reply biofoundation/model.py:27]: > model is no longer an AutoModelForCausalLM

🤦 

It would be nice if the auto class code enforced that your implementations have to extend the correct AutoClass types (e.g. I wonder why they didn't do that [here](https://huggingface.co/LongSafari/hyenadna-tiny-1k-seqlen-hf/blob/main/modeling_hyena.py#L401)?)

> I'm leaning towards (3)

👍

Comment (gonzalobenegas) [reply biofoundation/model.py:27]: I'm now running into the issue that the MLMs and CLMs must not only obey a protocol but probably also be a `nn.Module`, so we (currently the `Trainer`) can handle devices, compile, precision, etc. From my understanding this would require ABC, not Protocol. Alternatively, the protocol could also require `get_model(self) -> nn.Module` but this might be awkward.

Comment (eric-czech) [reply biofoundation/model.py:27]: Hmm, you could just do a runtime check on the nn.Module type?  Mypy follows `isinstance(obj, SomeType)` to narrow types anyhow, and it looks like there is a more explicit way to do this with a [TypeGuard](https://peps.python.org/pep-0647/).  I'm not sure that's worth it.

Maybe another way to go about it is:

- Assume you get an nn.Module instance
- Allow for an optional Callable that converts the nn.Module forward to the format you need
  - If not provided, you'd assume the model produces what you need already

Then you could let the module be the free-for-all of types and functions it usually is while still adding good typing to the function you need for conversion?

Comment (gonzalobenegas) [reply biofoundation/model.py:27]: Thanks, still trying to wrap my head around these options.

Option 1: an adapter class
Option 2: an adapter function

The advantage of option 1 is that the `run_inference` function doesn't need to keep track of the adapter function and just receive a single object.
The advantage of option 2 is that the user doesn't need to understand our class structure (which is a bit awkward as a mix of nn.Module and protocol/ABC).

Is that a good summary? I don't have developed a taste for this so if you think option 2 is currently better I'm happy to go with it.

File: biofoundation/model.py [hunk 3]
```diff
@@ -1,60 +1,115 @@
...
 
 
-class RefLogProbMLM(nn.Module):
-    """Reference Log Probability Masked Language Model wrapper.
+class HFLanguageModel(LanguageModel):
+    def __init__(self, model: PreTrainedModel):
+        super().__init__()
+        self.model = model
+
+    def forward(self, input_ids: Int[Tensor, "B L"]) -> Float[Tensor, "B L V"]:
```
[the remark this answers is not in the document]
Comment (gonzalobenegas) [reply biofoundation/model.py:26]: My understanding is that mypy only check if it's a `Tensor`, not the dtype or shape.

[the remark this answers is not in the document]
Comment (eric-czech) [reply biofoundation/model.py:26]: Ah, so all the dtype/shape checks happen at runtime?

File: biofoundation/model.py [hunk 4]
```diff
@@ -1,60 +1,135 @@
...
+    Computes the log-likelihoods using a softmax along the vocab dimension.
+    Uses the `input_ids` to index into the log-likelihoods and returns the likelihood
+    of the provided sequence at each position with dimension (batch, length-1).
+    """
+    softmax_logprobs = torch.log_softmax(logits, dim=-1)
+    softmax_logprobs = softmax_logprobs[:, :-1]
+    input_ids = input_ids[:, 1:]
+    assert softmax_logprobs.shape[1] == input_ids.shape[1]
+
+    logprobs = torch.gather(
```
[the remark this answers is not in the document]
Comment (gonzalobenegas) [reply biofoundation/model.py:119]: This makes sense to me. The model processes the first token and outputs the probability distribution for the second token. The model never outputs a probability distribution for the first token.

[the remark this answers is not in the document]
Comment (eric-czech) [reply biofoundation/model.py:119]: Ah, I see now 👍
